### PR TITLE
[PLATFORM-432] Add module search categories

### PR DESCRIPTION
**Pull request by tumppi**
Adds support for showing module category tree in the search. Categories can be toggled to show modules under it.

### CODE REVIEW

File: app/src/editor/canvas/components/ModuleSearch.jsx [hunk 1]
```diff
@@ -120,9 +177,91 @@ export class ModuleSearch extends React.PureComponent<Props, State> {
...
+            </React.Fragment>
+        ))
+    }
+
+    renderSearchResults = () => {
+        const { matchingModules, matchingStreams } = this.state
+        return (
+            <React.Fragment>
+                {matchingModules.length > 0 && (
+                    <div className={styles.SearchCategory}>Modules</div>
```
Comment (timoxley) [inline app/src/editor/canvas/components/ModuleSearch.jsx]: Probably shouldn't have hover effects/hand cursor for these non-interactive "Modules" and "Streams" header things:

image

Alternatively these could perhaps be treated as accordions that default to open, so if I'm specifically looking for a stream I can just close the accordion?

File: app/src/editor/canvas/components/ModuleSearch.jsx [hunk 2]
```diff
@@ -120,9 +177,91 @@ export class ModuleSearch extends React.PureComponent<Props, State> {
...
+
+        return modules
+    }
+
+    mapPathStringToArray = (pathString: string) => (
+        pathString.split(', ')
+    )
+
+    mapPathArrayToString = (pathArray: Array<string>) => (
+        pathArray.join(': ')
```
Comment (timoxley) [inline app/src/editor/canvas/components/ModuleSearch.jsx]: this joins with `:` but splits on `,`?

File: app/src/editor/canvas/state.js [hunk 3]
```diff
@@ -761,6 +761,7 @@ export function moduleTreeIndex(modules = [], path = [], index = []) {
                 id: m.metadata.id,
                 name: m.data,
                 path: path.join(', '),
+                pathArray: path,
```
Comment (timoxley) [inline app/src/editor/canvas/state.js]: IMO use one of `pathArray` or `path`, but not both.

File: app/src/editor/canvas/components/ModuleSearch.jsx [hunk 4]
```diff
@@ -10,7 +11,65 @@ import { moduleTreeSearch } from '../state'
...
+    toggle = () => {
+        this.setState(({ isExpanded }) => ({ isExpanded: !isExpanded }))
+    }
+
+    render() {
+        const { category, onItemSelect } = this.props
+        const { isExpanded } = this.state
+        return (
+            <React.Fragment>
+                {/* eslint-disable-next-line */}
```
Comment (juhah) [inline app/src/editor/canvas/components/ModuleSearch.jsx]: Would be good to specify the the excluded rule.

File: app/src/editor/canvas/components/ModuleSearch.jsx [hunk 5]
```diff
@@ -10,7 +11,65 @@ import { moduleTreeSearch } from '../state'
...
+                {isExpanded && category.modules.map((m) => (
+                    <ModuleMenuItem key={m.id} module={m} onSelect={onItemSelect} />
+                ))}
+            </React.Fragment>
+        )
+    }
+}
+
+const ModuleMenuItem = ({ module, onSelect }) => (
+    /* eslint-disable-next-line */
```
Comment (juhah) [inline app/src/editor/canvas/components/ModuleSearch.jsx]: Same here

File: app/src/editor/canvas/components/ModuleSearch.jsx [hunk 6]
```diff
@@ -120,9 +177,91 @@ export class ModuleSearch extends React.PureComponent<Props, State> {
         }
     }
 
+    getMappedModuleTree = (search: string) => {
```
Comment (juhah) [inline app/src/editor/canvas/components/ModuleSearch.jsx]: Use default value `search: string = ''`?

**Pull request merged**